File: woodroid/src/main/java/me/gilo/woodroid/data/api/CustomerAPI.java
package me.gilo.woodroid.data.api;


import me.gilo.woodroid.models.Coupon;
import me.gilo.woodroid.models.Customer;
import me.gilo.woodroid.models.Download;
import retrofit2.Call;
import retrofit2.http.*;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public interface CustomerAPI {

    @Headers("Content-Type: application/json")
    @POST("customers")
    Call<Customer> create(@Body Customer body);

    @GET("customers/{id}")
    Call<Customer> view(@Path("id") int id);

    @GET("customers")
    Call<List<Customer>> list();

    @Headers("Content-Type: application/json")
    @PUT("customers/{id}")
    Call<Customer> update(@Path("id") int id, @Body Customer body);

    @DELETE("customers/{id}")
    Call<Customer> delete(@Path("id") int id);

    @DELETE("customers/{id}")
    Call<Customer> delete(@Path("id") int id, @Query("force") boolean force);

    @POST("customers/batch")
    Call<String> batch(@Body Customer body);

    @POST("customers/{id}/downloads")
    Call<List<Download>> downloads(@Path("id") int id);

    @GET("customers")
    Call<List<Customer>> filter(@QueryMap Map<String, String> filter);

}

File: woodroid/src/main/java/me/gilo/woodroid/data/api/CustomerAPI.kt
package me.gilo.woodroid.data.api


import me.gilo.woodroid.models.Customer
import me.gilo.woodroid.models.Download
import retrofit2.Call
import retrofit2.http.*

interface CustomerAPI {

    @Headers("Content-Type: application/json")
    @POST("customers")
    fun create(@Body body: Customer): Call<Customer>

    @GET("customers/{id}")
    fun view(@Path("id") id: Int): Call<Customer>

    @GET("customers")
    fun list(): Call<List<Customer>>

    @Headers("Content-Type: application/json")
    @PUT("customers/{id}")
    fun update(@Path("id") id: Int, @Body body: Customer): Call<Customer>

    @DELETE("customers/{id}")
    fun delete(@Path("id") id: Int): Call<Customer>

    @DELETE("customers/{id}")
    fun delete(@Path("id") id: Int, @Query("force") force: Boolean): Call<Customer>

    @POST("customers/batch")
    fun batch(@Body body: Customer): Call<String>

    @POST("customers/{id}/downloads")
    fun downloads(@Path("id") id: Int): Call<List<Download>>

    @GET("customers")
    fun filter(@QueryMap filter: Map<String, String>): Call<List<Customer>>

}